lib/config.js

```diff
@@ -6,17 +6,104 @@
  * Reads and validates the master and worker config.json files.
  */
 
+'use strict';
+
+
 var fs = require('fs-extra');
+var logger = require('./logger')
 var path = require('path');
 var url = require('url');
-var argv = require('yargs').argv;
+var argv = require('yargs')
+    .usage('Usage: $0 {--master | --worker} [--config ConfigFile]')
+	.boolean('master')
+	.boolean('worker')
+	.default('config', false, 'Optional config file location. Defaults to {guerillaDir}/config')
+	.argv;
+
+if (argv.config && argv.config === true) {
+	throw new Error("--config defined without any following configFile")
+}
+
+var mode = 'unknown'
+
+var expandTilde = function(apath) {
+	var homedir = process.env['HOME'];
+	if (!apath) return apath;
+	if (apath == '~') return homedir;
+	if (apath.slice(0, 2) != '~/') return apath;
+	return path.join(homedir, apath.slice(2));
+}
+
+var initMode = function () {
+	if (argv.master && argv.worker) throw new Error("Cannot invoke with both --master and --worker");
+	if (argv.master) {
+		mode = 'master';
+		return;
+	}
+	if (argv.worker) {
+		mode = 'worker';
+		return;
+	}
+	throw new Error('Invalid mode.');
+}
+
+var getConfig = function(mode) {
+	var pathArray = [];
+	if(argv.config ) {
+		var altConfigPath = expandTilde(argv.config);
+		altConfigPath = path.resolve(altConfigPath);
+		logger.i('--config command line flag detected. Config is from: ' + altConfigPath);
+		pathArray.push(altConfigPath);
+	}
+	else {
+		pathArray.push(__rootdir, 'config', mode, 'config.json');
+	}
+	return require(path.join.apply(this, pathArray))
+}
```

Comment:
path.join accepts only string arguments per my reading. I have an array that I need to apply to the function.